midp/gfx.js

```diff
@@ -1178,38 +1135,114 @@ var currentlyFocusedTextEditor;
     var TRANS_ROT270 = 6;
     var TRANS_MIRROR_ROT90 = 7;
 
-    function renderRegion(g, image, sx, sy, sw, sh, transform, x, y, anchor) {
-        var imgData = image.imageData,
-            texture = imgData.context.canvas;
-
-        var c = g.context2D;
-        if (transform !== TRANS_NONE) {
-            c.save();
-        }
-
-        var pair = withAnchor(g, c, anchor, x, y, sw, sh);
-        x = pair[0];
-        y = pair[1];
-
-        if (transform === TRANS_MIRROR || transform === TRANS_MIRROR_ROT180) {
-            c.scale(-1, 1);
-        } else if (transform === TRANS_MIRROR_ROT90 || transform === TRANS_MIRROR_ROT270) {
-            c.scale(1, -1);
-        } else if (transform === TRANS_ROT90 || transform === TRANS_MIRROR_ROT90) {
-            c.rotate(Math.PI / 2);
-        } else if (transform === TRANS_ROT180 || transform === TRANS_MIRROR_ROT180) {
-            c.rotate(Math.PI);
-        } else if (transform === TRANS_ROT270 || transform === TRANS_MIRROR_ROT270) {
-            c.rotate(1.5 * Math.PI);
+    function renderRegion(dstContext, srcCanvas, sx, sy, sw, sh, transform, absX, absY, anchor) {
+        var w, h;
+        switch (transform) {
+            case TRANS_NONE:
+            case TRANS_ROT180:
+            case TRANS_MIRROR:
+            case TRANS_MIRROR_ROT180:
+                w = sw;
+                h = sh;
+                break;
+            case TRANS_ROT90:
+            case TRANS_ROT270:
+            case TRANS_MIRROR_ROT90:
+            case TRANS_MIRROR_ROT270:
+                w = sh;
+                h = sw;
+                break;
+        }
+
+        // Make `absX` and `absY` the top-left coordinates where we will
+        // place the image in absolute coordinates
+        if (0 != (anchor & HCENTER)) {
+            absX -= (w / 2);
```

Comment:
Can you use *(w >>> 1) | 0*? Maybe it's the absence of truncation that is generating the small differences in the rendering?